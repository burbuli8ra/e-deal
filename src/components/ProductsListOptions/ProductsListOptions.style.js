import styled from '@emotion/styled';
import { colors } from 'theme';

const FilterLabel = styled.div`
  color: ${colors.pastelBlue};
  font-size: 14px;
  font-weight: 700;
  margin: 0 4px;
`;
FilterLabel.displayName = 'FilterLabel';

const Filters = styled.div`
  align-items: center;
  display: flex;
`;
Filters.displayName = 'Filters';

const Label = styled.div`
  color: ${colors.pastelBlue};
  font-size: 14px;
  font-weight: 700;
  margin-bottom: 4px;
`;
Label.displayName = 'Label';

const OptionColumn = styled.div``;

const Options = styled.div`
  background-color: ${colors.white};
  border-radius: 4px;
  box-shadow: 0 0 9px 0 rgba(140, 140, 158, 0.25);
  column-gap: 20px;
  row-gap: 20px;
  display: grid;
  grid-template-columns: 2fr 1fr 1fr;
  grid-template-rows: 100%;
  flex: 0 1 100%;
  margin-bottom: 10px;
  padding: 20px;
  
  @media (max-width: 520px) {
    grid: initial;
  }
`;
Options.displayName = 'Options';

const OrderOptions = styled.div`
  display: flex;
  flex-direction: column;
  
  > * {
    margin-bottom: 4px;
    
    &:last-of-type {
      margin-bottom: 0;
    }
  }
`;
OrderOptions.displayName = 'OrderOptions';

export default {
  FilterLabel,
  Filters,
  Label,
  OptionColumn,
  Options,
  OrderOptions
};