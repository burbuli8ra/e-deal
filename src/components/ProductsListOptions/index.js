export { default as ProductsListOptions } from './ProductsListOptions';