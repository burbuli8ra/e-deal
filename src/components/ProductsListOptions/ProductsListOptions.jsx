import React, { useState } from 'react';
import { RadioButton, TextInput } from 'components';
import Styled from './ProductsListOptions.style';

const KeyCodes = {
  ENTER: 13
};

const ProductsListOptions = ({
  categoryMaxPrice,
  categoryMinPrice,
  onMaxPriceUpdate,
  onMinPriceUpdate,
  onOrderByUpdate,
  onOrderTypeUpdate,
  orderByOptions,
  orderByValue,
  orderTypeOptions,
  orderTypeValue
}) => {
  const [maxPrice, setMaxPrice] = useState('');
  const [minPrice, setMinPrice] = useState('');

  const handlePriceChange = (fieldType, { target: { value } }) => {
    // Allow only numbers and empty
    const regex = /^(\s*|\d+)$/;

    if (!regex.test(value)) {
      return;
    }

    if (fieldType === 'max') {
      setMaxPrice(value);
      return;
    }

    setMinPrice(value);
  };

  const handleMaxPriceUpdate = (
    getPriceInCents,
    valueInCents,
    maxPriceInEuros,
    minPriceInEuros
  ) => {
    // Case of greater input value than max limit
    if (valueInCents > categoryMaxPrice) {
      setMaxPrice(maxPriceInEuros.toString());
      onMaxPriceUpdate(categoryMaxPrice);
      return;
    }

    // Case of lower input value than min limit
    if (valueInCents < categoryMinPrice) {
      setMaxPrice((minPriceInEuros + 1).toString());
      onMaxPriceUpdate(categoryMinPrice + 100);
      return;
    }

    // Case of lower input value than value of min field
    if (valueInCents < getPriceInCents(minPrice)) {
      setMaxPrice((parseInt(minPrice ,10) + 1).toString());
      onMaxPriceUpdate(getPriceInCents(minPrice) + 100);
      return;
    }

    onMaxPriceUpdate(valueInCents);
  };

  const handleMinPriceUpdate = (
    getPriceInCents,
    valueInCents,
    maxPriceInEuros,
    minPriceInEuros
  ) => {
    // Case of greater input value than max limit
    if (valueInCents > categoryMaxPrice) {
      setMinPrice((maxPriceInEuros - 1).toString());
      onMinPriceUpdate(categoryMaxPrice - 100);
      return;
    }

    // Case of lower input value than min limit
    if (valueInCents < categoryMinPrice) {
      setMinPrice(minPriceInEuros.toString());
      onMinPriceUpdate(categoryMinPrice);
      return;
    }

    // Case of greater input value than value of max field
    if (valueInCents > getPriceInCents(maxPrice)) {
      setMinPrice((parseInt(maxPrice ,10) - 1).toString());
      onMinPriceUpdate(getPriceInCents(maxPrice) - 100);
      return;
    }

    onMinPriceUpdate(valueInCents);
  };

  const handlePriceUpdate = (fieldType, { target: { value } }) => {
    const getPriceInCents = price => parseInt(price, 10) * 100;
    const valueInCents = getPriceInCents(value);
    const maxPriceInEuros = Math.ceil(categoryMaxPrice/100);
    const minPriceInEuros = Math.floor(categoryMinPrice/100);

    if (fieldType === 'max') {
      handleMaxPriceUpdate(
        getPriceInCents,
        valueInCents,
        maxPriceInEuros,
        minPriceInEuros
      );
    }

    if (fieldType === 'min') {
      handleMinPriceUpdate(
        getPriceInCents,
        valueInCents,
        maxPriceInEuros,
        minPriceInEuros
      );
    }
  };

  const handlePriceKeyPress = (fieldType, e) => {
    const { charCode } = e;

    if (charCode === KeyCodes.ENTER) {
      handlePriceUpdate(fieldType, e);
    }
  };

  return (
    <Styled.Options>
      <Styled.OptionColumn>
        <Styled.Label>Filter by Price (€)</Styled.Label>
        <Styled.Filters>
          <TextInput
            onBlur={e => handlePriceUpdate('min', e)}
            onChange={e => handlePriceChange('min', e)}
            onKeyPress={e => handlePriceKeyPress('min', e)}
            placeholder="Min €"
            value={minPrice}
          />
          <Styled.FilterLabel>to</Styled.FilterLabel>
          <TextInput
            onBlur={e => handlePriceUpdate('max', e)}
            onChange={e => handlePriceChange('max', e)}
            onKeypress={e => handlePriceKeyPress('max', e)}
            placeholder="Max €"
            value={maxPrice}
          />
        </Styled.Filters>
      </Styled.OptionColumn>
      <Styled.OptionColumn>
        <Styled.Label>Order by</Styled.Label>
        <Styled.OrderOptions>
          {orderByOptions.map(item => (
            <RadioButton
              key={item.value}
              label={item.value}
              value={orderByValue}
              onClick={() => onOrderByUpdate(item.value)}
            >
              {item.label}
            </RadioButton>
          ))}
        </Styled.OrderOptions>
      </Styled.OptionColumn>
      <Styled.OptionColumn>
        <Styled.Label>Order Type</Styled.Label>
        <Styled.OrderOptions>
          {orderTypeOptions.map(item => (
            <RadioButton
              key={item.value}
              label={item.value}
              value={orderTypeValue}
              onClick={() => onOrderTypeUpdate(item.value)}
            >
              {item.label}
            </RadioButton>
          ))}
        </Styled.OrderOptions>
      </Styled.OptionColumn>
    </Styled.Options>
  );
};

export default ProductsListOptions;