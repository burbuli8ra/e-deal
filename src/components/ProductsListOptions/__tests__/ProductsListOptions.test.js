import React from 'react';
import { render } from '@testing-library/react';
import ProductsListOptions from '../ProductsListOptions';

const setup = props => <ProductsListOptions {...props} />;

describe('ProductsListOptions component', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      categoryMaxPrice: 10000,
      categoryMinPrice: 1000,
      onMaxPriceUpdate: jest.fn(),
      onMinPriceUpdate: jest.fn(),
      onOrderByUpdate: jest.fn(),
      onOrderTypeUpdate: jest.fn(),
      orderByOptions: [
        { label: 'price', value: 'Price' },
        { label: 'title', value: 'Title' }
      ],
      orderByValue: 'price',
      orderTypeOptions: [
        { label: 'asc', value: 'Ascending' },
        { label: 'dec', value: 'Descending' }
      ],
      orderTypeValue: 'asc'
    };
    wrapper = render(setup(props));
  });

  test('should match snapshot', () => {
    const { asFragment } = wrapper;
    expect(asFragment()).toMatchSnapshot();
  });
});