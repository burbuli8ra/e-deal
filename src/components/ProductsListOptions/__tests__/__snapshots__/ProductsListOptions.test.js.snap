// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsListOptions component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1qn0r3y"
  >
    <div
      class="css-0"
    >
      <div
        class="css-1ux9f3u"
      >
        Filter by Price (€)
      </div>
      <div
        class="css-112jqer"
      >
        <input
          class="css-1vb9rdv"
          data-testid="text-input"
          placeholder="Min €"
          type="text"
          value=""
        />
        <div
          class="css-yewzub"
        >
          to
        </div>
        <input
          class="css-1vb9rdv"
          data-testid="text-input"
          placeholder="Max €"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="css-0"
    >
      <div
        class="css-1ux9f3u"
      >
        Order by
      </div>
      <div
        class="css-1skiy8z"
      >
        <div
          class="css-19cy8ai-RadioButton e1xvq25d1"
        >
          <div
            class="css-10vhthi-Bullet e1xvq25d0"
          />
          price
        </div>
        <div
          class="css-19cy8ai-RadioButton e1xvq25d1"
        >
          <div
            class="css-10vhthi-Bullet e1xvq25d0"
          />
          title
        </div>
      </div>
    </div>
    <div
      class="css-0"
    >
      <div
        class="css-1ux9f3u"
      >
        Order Type
      </div>
      <div
        class="css-1skiy8z"
      >
        <div
          class="css-19cy8ai-RadioButton e1xvq25d1"
        >
          <div
            class="css-10vhthi-Bullet e1xvq25d0"
          />
          asc
        </div>
        <div
          class="css-19cy8ai-RadioButton e1xvq25d1"
        >
          <div
            class="css-10vhthi-Bullet e1xvq25d0"
          />
          dec
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
